### ISSUE TITLE
Support for Meteor

### ISSUE THREAD

**Issue by Sanjo** [wallabyjs/public#215]
Hi, I'm one of the main contributors to [Velocity](https://github.com/meteor-velocity/velocity), the testing solution for [Meteor](https://www.meteor.com/). I'd like to bring Wallaby to Meteor because of the great feedback cycle that Wallaby offers. I have already successfully integrated Jasmine and the Karma Runner with Meteor before.

I use this issue to track my progress with integrating Wallaby and Meteor. I will also ask questions if I encounter problems that I cannot solve myself.

I have read the documentation on the different parts of Wallaby and are already using Wallaby for a node.js package.

I have created a very simple Meteor app [as example here](https://github.com/Sanjo/meteor-wallaby-sandbox) that I will use as reference to describe and solve the problems I encounter while doing the integration.

__Status:__
* Simple server tests with Mocha work

__TODO:__
* [ ] Support browser tests
* [ ] Support parallelization
* [ ] Support reusing instances (recycle: false)

__Open question:__
* How can I configure Wallaby for server and client at the same time?

**Comment by ArtemGovorov** [comment]
Awesome! I haven't used Meteor extensively and only briefly know the way it works, but more than happy to help making wallaby work with it, including wallaby core changes if required. 

How stable is the current solution for node.js tests? I'd like to add it to wallaby recipes in docs or you reckon it's too early?

> How can I configure Wallaby for server and client at the same time?

For client tests you need to create another configuration file and run it as a separate Run Configuration in IDE. 

There's also a [limitation at the moment](https://github.com/wallabyjs/public/issues/89), that only one wallaby configuration can be running at the same time for an opened project (if you can open your server and client in 2 separate IDE instances, then it should work). I'll need to tackle code coverage results merging and a few other things on the IDE side to make it work. I was also thinking of extending the config file to allow configuring both client and server tests in one config, so that some bits can be easier re-used (though they can be re-used even with multiple configs).

I suggest to try make it work separately for know. Making them work together will requires some wallaby core and IDE plugin work.

> Support browser tests

If Karma works for Meteor, it definitely should be possible to configure wallaby as well.

> Support parallelization

As mentioned in your config file comments, random ports are required if a real web server is spinned up. [Node.js http server](https://nodejs.org/api/http.html#http_server_listen_path_callback) just uses a random port if the port is not specified when calling `listen`, I'm assuming Meteor is using node http server behind the scenes and maybe allow not passing the port as well?

Some DB setup may be required as well, such as possibly creating separate databases per worker. Have a look into this [issue for more details](https://github.com/wallabyjs/public/issues/171).

**Comment by Sanjo** [comment]
> How stable is the current solution for node.js tests? I'd like to add it to wallaby recipes in docs or you reckon it's too early?

I will add more tests tomorrow. I will let you know if I find any other issues.

> I suggest to try make it work separately for know. Making them work together will requires some wallaby core and IDE plugin work.

I will make it work separately for know. Meteor apps usually have a lot of isomorphic code and server and client code is not strictly separated. So having support to run Wallaby a node.js server configuration and a Karma client configuration at the same time for one project would be great.

> As mentioned in your config file comments, random ports are required if a real web server is spinned up. Node.js http server just uses a random port if the port is not specified when calling listen, I'm assuming Meteor is using node http server behind the scenes and maybe allow not passing the port as well?

> Some DB setup may be required as well, such as possibly creating separate databases per worker. Have a look into this issue for more details.

I will try that tomorrow too. You can pass the MONGO_URL and PORT as environment variables. So should theoretically be no problem if that are the only problems with parallel running.

**Comment by ArtemGovorov** [comment]
> So having support to run Wallaby a node.js server configuration and a Karma client configuration at the same time for one project would be great.

Do you mean Wallaby client configuration? Because running Wallaby node config and Karma client at the same time for the same project should already work, wallaby doesn't conflict with Karma.

**Comment by Sanjo** [comment]
> Do you mean Wallaby client configuration?

Yes. I thought Wallaby works with Karma because I read Karma somewhere. Haven't used Wallaby for client code yet.

**Comment by Sanjo** [comment]
I've added some more tests and found some problems:

For all reproduces you need to run [the Meteor app](https://github.com/xolvio/meteor-wallaby-sandbox) with the command `meteor` (You need to have [Meteor installed](https://www.meteor.com/install)).
You need to start a local MongoDB server.
You also need to start Wallaby with the .wallaby.js config file.

#### 1.) WebStorm does not detect changes of built files:

__Reproduce:__

In the file `meteor-wallaby-sandbox.js` change `'getArticles'` to `'getArticles2'` and save.

__Expected:__ Wallaby detects change and reruns the test `tests/wallaby/methods/getArticlesSpec.js`. 

__What happens:__ Webstorm does not recognize the change in the built file `.meteor/local/build/programs/server/app/meteor-wallaby-sandbox.js` until you explicitly synchronize the file.

**Comment by ArtemGovorov** [comment]
It looks like WebStorm/IntelliJ platform doesn't detect the change or doesn't notify about it. Wallaby.js plugin is subscribed to all Virtual File System events from the IDE, and in this case it's not receiving one.

Does Meteor have an API/node module to build/incrementally rebuild a project? I was thinking that perhaps a better idea than watching built files would be to watch original source files, `meteor-wallaby-sandbox.js` in this case. Otherwise, using the current approach, original source files are not tracked, the code coverage is not displayed for them but only for test files, etc.

If Meteor supports some API that can be invoked from node.js and is capable of generating source maps (between original and generated files), then perhaps it's better to implement the support as a wallaby [postprocessor](http://wallabyjs.com/docs/config/postprocessor.html), like we do for [Webpack](https://github.com/jeffling/wallaby-webpack/blob/master/index.js) and [Browserify](https://github.com/wallabyjs/wallabify/blob/master/index.js). In this postprocessor, Meteor could be started, notified about file changes and told to rebuild stuff, etc. Generated source maps would allow user to get the code coverage and inline messages in the original source files.

**Comment by Sanjo** [comment]
I have to look into the Meteor build system code. Should be doable, but I have to learn it first.

**Comment by faceyspacey** [comment]
Not sure if this information is helpful, but basically Meteor launched a new Build Plugin API since this issue was last active. I've used it extensively. Essentially, all the files in the app are passed to any specified "compiler build plugins" (currently, it only supports one per file extension, but you can manually combine the code from multiple extensions into one, as I've done before), and here you can process the files and output them to the same or a new file name. This is where babel is hooked. It's pretty straightforward, and it can generate source maps.  

The resulting bundle is stored in a place you can then easily access (`.meteor/local/build/programs/server` and `.meteor/local/build/programs/web.browser`), with source maps and all. 

Testing has been a soar spot for the Meteor community. This could be big for us, especially with a lot of the React testing tools evolving and starting to solve our problems as Meteor moves from their in-house reactive view framework, *Blaze*, to React. 

https://github.com/meteor/meteor/wiki/Build-Plugins-API

**Comment by tomitrescak** [comment]
This would be nothing less but awesome having Meteor work with Wallaby!

**Comment by Sanjo** [comment]
I think it needs to be implemented as Artem described in his last comment (https://github.com/wallabyjs/public/issues/215#issuecomment-125100820).

Or we wait until Meteor is based on Webpack.

**Comment by ArtemGovorov** [comment]
> Or we wait until Meteor is based on Webpack.

Oh, is there such a plan?

**Comment by faceyspacey** [comment]
@Sanjo 
>I think it needs to be implemented as Artem described in his last comment 

Are you referring to the rebuild "incrementally" part? Does it really have to rely on being rebuilt incrementally--couldn't we just have a slower version of Wallaby? What exactly would he need to implement this in your opinion? 

because it seems we have this solved:
>If Meteor supports some API that can be invoked from node.js and is capable of generating source maps (between original and generated files), then perhaps it's better to implement the support as a wallaby postprocessor, like we do for Webpack and Browserify. In this postprocessor, Meteor could be started, notified about file changes and told to rebuild stuff, etc. Generated source maps would allow user to get the code coverage and inline messages in the original source files.

**Comment by Sanjo** [comment]
@faceyspacey 

I didn't refer to any specific aspect. But yes, the rebuilt needs to be incremental, otherwise it is to slow. Maybe we can just let Meteor do its usual building and the Wallaby postprocessor would just make Wallaby aware of the built files from .meteor/local/build.

Regarding your earlier post about the Build Plugin API: I don't see how this API helps with this problem. Maybe you don't have the full understanding of how Wallaby works or I miss something.

@ArtemGovorov It's likely that they will base their build tool on Webpack in a future Meteor 2, based on the [research project](https://forums.meteor.com/t/sharing-some-research-into-webpack-and-the-build-tool-ecosystem/12239) they have done and some other comments. But that's probably not coming in the next 3 months. But there are already approaches to use Webpack with Meteor from the community.

**Comment by faceyspacey** [comment]
@Sanjo 
> Maybe we can just let Meteor do its usual building and the Wallaby postprocessor would just make Wallaby aware of the built files from .meteor/local/build.

That would be the perfect first step, and would essentially work. I mean at that point basically Wallaby has done its job, and it's on Meteor to get building faster. It's already quite a bit faster. For small projects, it's super fast these days. I don't see why we wanna stop this integration because it's not as fast as other Wallaby integrations. 

That said, we may have ways to get it faster on our side: we already have webpack integration via several techniques. I.e. within Meteor like @eXon built or, of if you use Meteor as an NPM package like @AdamBrodzinski built. The latter might already allow for Wallaby usage. 

So anyway, if we could get wallaby's side done, we can find ways to make the build process quicker--perhaps doing similar techniques that @eXon did to hi-jack the build process and just rebuild the tests, for example. 

**If a big part of the job is telling Wallaby when the build files are complete and where they are, I can provide them a package that does that any time. Let me know.**

**Comment by tomitrescak** [comment]
I'm working on integrating wallaby with @jedwards and @AdamBrodzinski's solution. Once I fully understand what is needed there, it should be possible to do as such for @eXon's solution as well.

The @jedwards solution should be fairly strightforward for unit tests, since he's building a single file for client and server. How to do integration tests is however big unknown to me.

**Comment by lucidprogrammer** [comment]
Any luck @tomitrescak ?

**Comment by tomitrescak** [comment]
@lucidprogrammer check out Mantra framework for Meteor. It has full support for Wallaby.js.

**Comment by lucidprogrammer** [comment]
@tomitrescak yes I am playing with the mantra kickstarter. However, I have a fair bit of server code which currently using velocity in integration mode. (mantra seems to focus on the client only) With the changes coming through, I was looking for options.

**Comment by Sanjo** [comment]
We have this working for modules only apps. See https://github.com/xolvio/automated-testing-best-practices/tree/wip/migrate_1.3.

To make it work with apps that don't just use modules, you need to adjust the file loading patterns.

**Comment by lucidprogrammer** [comment]
that works perfectly well, great work sanjo
